bob-access/src/authenticator/basic.rs

```diff
@@ -1,65 +1,132 @@
+    fn spawn_resolver(&self, cred: DeclaredCredentials) {
+        tokio::spawn(Self::resolve_worker(self.nodes.clone(), cred, self.resolve_sleep_period_ms));
+    }
+
+    async fn resolve_worker(nodes: Arc<RwLock<NodesCredentials>>, cred: DeclaredCredentials, sleep_period_ms: u64) {
+        let hostname = cred.hostname().as_ref().expect("resolve worker without hostname");
+        let mut addr: Option<Vec<SocketAddr>> = lookup(hostname).await;
+        let mut cur_sleep_period_ms = 100;
+        while addr.is_none() || (addr.is_some() && addr.as_ref().unwrap().len() == 0) {
+            tokio::time::sleep(Duration::from_millis(cur_sleep_period_ms)).await;
+
+            addr = lookup(hostname).await;
+
+            cur_sleep_period_ms = sleep_period_ms.min(cur_sleep_period_ms * 2);
+        }
+
+        let addr = addr.expect("somehow addr is none");
+        if let CredentialsKind::InterNode(nodename) = cred.kind() {
+            let mut nodes = nodes.write().expect("nodes credentials lock");
+            if let Some(creds) = nodes.get_mut(nodename) {
+                creds.set_resolved(addr);
+            }
+        } else {
+            error!("resolved credentials are not internode");
+        }
+    }
+
+    fn process_auth_result(&self, nodename: &str, authenticated: bool) {
+        let mut nodes = self.nodes.write().expect("nodes credentials lock");
+        if let Some(node) = nodes.get_mut(nodename) {
+            if node.update_resolve_state(authenticated) {
+                node.set_in_progress();
+                self.spawn_resolver(node.creds().clone());
+            }
         }
-        self.nodes
-            .get(ip.as_ref()?)
-            .map(|creds| {
-                for cred in creds {
-                    if let Some(CredentialsKind::InterNode(other_name)) = cred.kind() {
-                        if node_name == other_name {
-                            return true;
-                        }
+    }
+
+    fn check_node_request(&self, node_name: &String, ip: Option<SocketAddr>) -> bool {
+        let mut authenticated = None;
+        {
+            if ip.is_none() {
+                return false;
+            }
+            let ip = ip.unwrap().ip();
+            let nodes = self.nodes.read().expect("nodes credentials lock");
+            if let Some(cred) = nodes.get(node_name).map(|guard| guard.creds()) {
+                if let CredentialsKind::InterNode(other_name) = cred.kind() {
+                    debug_assert!(node_name == other_name);
+                    if cred.ip().iter().find(|cred_ip| cred_ip.ip() == ip).is_some() {
+                        authenticated = Some(true);
+                    } else {
+                        authenticated = Some(false);
                     }
                 }
-                false
-            })
+            }
+        }
+        if let Some(authenticated) = authenticated {
+            self.process_auth_result(node_name, authenticated);
```

Comment:
`process_auth_result` acquires write lock inside, so it is better to avoid calling it in most common scenarios. I suggest to add a function to `DCredentialsResolveGuard` that checks whether the status update needed (it is not needed if  `resolve_state == Resolved(0) && authenticated` or `resolve_state == InProgress`). Basically the logic is the same as in `update_resolve_state`, but without the actual update. Then you can call it inside read lock above and if it returned `false` then do not call `process_auth_result`.